src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -201,8 +225,17 @@ Object* FrozenObjectSegment::TryAllocateObject(PTR_MethodTable type, size_t obje
 
     m_pCurrent += objectSize;
 
-    // Notify GC that we bumped the pointer and, probably, committed more memory in the reserved part
-    GCHeapUtilities::GetGCHeap()->UpdateFrozenSegment(m_SegmentHandle, m_pCurrent, m_pStart + m_SizeCommitted);
+    if (IsRegistered())
+    {
+        // Notify GC that we bumped the pointer and, probably, committed more memory in the reserved part
+        // NOTE: UpdateFrozenSegment is not expected to take any lock inside
```

Comment:
This sounds a bit suspicious. What guarantees that some other GC thread is not reading the limits as they are being updated?

Reply:
Do you mean it should take the lock in GC and FrozenManager should take the registration lock for every allocation?

Reply:
I do not know how and where the limits are used by the GC. It just looks suspicious that we do not need any synchronization when updating them.

Reply:
My guess that generally GC just ignores ro segments except:
1) scanning in-range ro-segements in segments-based GC (e.g. 32bit)
2) reporting them to debugger via EnumMemory

so probably it does need a lock (but am not sure). maybe @Maoni0 @cshung can advise? 
Question is:
Does 
https://github.com/dotnet/runtime/blob/899461780a9e309142e8d85f25e15d4416157fa7/src/coreclr/gc/gcee.cpp#L510-L517

need to take `gc_heap::gc_lock` lock?

Reply:
Added locking via https://github.com/dotnet/runtime/pull/90847/commits/9c944ecfbcf012575cd1df110b9a66ded4a51fa8 but still would be nice to hear an opinion from GC team

Reply:
Looks like it's definitely needed, otherwise Profiler will have a race condition

Reply:
@cshung, can you please take a look at this?

Reply:
I am on it.